src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RSA.cs

```diff
@@ -980,6 +1362,79 @@ private static void ClearPrivateParameters(in RSAParameters rsaParameters)
             CryptographicOperations.ZeroMemory(rsaParameters.InverseQ);
         }
 
+        private delegate bool TryFunc<TState>(RSA rsa, ReadOnlySpan<byte> input, byte[] destination, TState state, out int bytesWritten);
+
+        private byte[] TryWithKeyBuffer<TState>(
+            ReadOnlySpan<byte> input,
+            TState state,
+            TryFunc<TState> callback,
+            bool tryKeySizeFirst = true)
+        {
+            // In normal circumstances, the signing and encryption size is the key size.
+            // In the case of decryption, it will be at most the size of the key, but the final output size is not
+            // deterministic, so start with the key size.
+            int resultSize = (KeySize + 7) / 8;
+            int written;
+
+            // For scenarios where we are confident that we can get the output side right on the first try, we allocate
+            // and use that as the buffer. This is the case for signing and encryption since that is always going to be
+            // the modulus size.
+            // For decryption, we go straight to renting as there is no way to know the size of the final output prior
+            // to decryption, so the allocation would probably be wasted.
+            if (tryKeySizeFirst)
+            {
+                byte[] result = new byte[resultSize];
+
+                if (callback(this, input, result, state, out written))
+                {
+                    if (written <= resultSize)
+                    {
+                        // In a typical sign or encrypt scenario, Resize won't do anything so we return the array as-is.
+                        // On the offchance some implementation returns less than what we expected, shrink the result.
+                        Array.Resize(ref result, written);
+                        return result;
+                    }
+
+                    // The Try virtual returned a bogus value for written. It returned a written value larger
+                    // than the buffer passed in. This means the Try implementation is not reliable, so throw
+                    // the best exception we can.
+                    throw new CryptographicException(SR.Argument_DestinationTooShort);
+                }
+
+                // We're about to try renting from the pool, so the next rental should be bigger than what we just tried.
+                resultSize = checked(resultSize * 2);
```

Comment:
Do we really have a growth scenario here?  (Does this line stand a chance of being covered?)

Reply:
In the real world, no. We should never get anything back bigger than the modulus (key) size.

But, we depend on a virtual. Someone could be doing some "clever" so I figured we can at least make it work. The happy-path works and we just allocate and return. For the unhappy path, it works. The happy path isn't penalized, the only penalty we are paying is a little bit of complexity.

If we want to just throw if the virtual gives us something back that is not "RSA", we can. It would certainly allow me to delete some tests.

Reply:
Here's a test for the growth scenario:

https://github.com/dotnet/runtime/blob/5167e3498f36f844f661186456f459340153d6b8/src/libraries/System.Security.Cryptography/tests/RSATests.cs#L503

Reply:
Ah, yes. And our base TryDecrypt just trusts the virtual, so this is being consistent.